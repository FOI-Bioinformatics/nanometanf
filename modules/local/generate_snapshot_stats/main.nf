process GENERATE_SNAPSHOT_STATS {
    tag "$batch_meta.batch_id"
    label 'process_single'
    publishDir "${params.outdir}/realtime_stats/snapshots", mode: 'copy'

    input:
    tuple val(batch_meta), val(file_metas)
    val stats_config

    output:
    tuple val(batch_meta), path("${batch_meta.batch_id}_snapshot.json"), emit: snapshot_stats
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    
    """
    #!/usr/bin/env python3
    
    import json
    import time
    from datetime import datetime
    from pathlib import Path
    
    # Batch metadata
    batch_meta = ${groovy.json.JsonBuilder(batch_meta).toString()}
    file_metas = ${groovy.json.JsonBuilder(file_metas).toString()}
    stats_config = ${groovy.json.JsonBuilder(stats_config).toString()}
    
    # Calculate snapshot statistics
    snapshot_stats = {
        'batch_info': {
            'batch_id': batch_meta['batch_id'],
            'batch_timestamp': batch_meta['batch_timestamp'],
            'batch_time_formatted': batch_meta['batch_time'],
            'processing_timestamp': int(time.time() * 1000),
            'processing_time_formatted': datetime.now().isoformat()
        },
        'file_statistics': {
            'file_count': len(file_metas),
            'total_size_bytes': sum(f.get('file_size', 0) for f in file_metas),
            'total_size_mb': round(sum(f.get('file_size', 0) for f in file_metas) / 1024 / 1024, 2),
            'estimated_total_reads': sum(f.get('estimated_reads', 0) for f in file_metas),
            'compressed_files': sum(1 for f in file_metas if f.get('is_compressed', False)),
            'average_file_size_mb': round((sum(f.get('file_size', 0) for f in file_metas) / len(file_metas)) / 1024 / 1024, 2) if file_metas else 0
        },
        'priority_analysis': {
            'average_priority': round(sum(f.get('priority_score', 0) for f in file_metas) / len(file_metas), 2) if file_metas else 0,
            'max_priority': max((f.get('priority_score', 0) for f in file_metas), default=0),
            'min_priority': min((f.get('priority_score', 0) for f in file_metas), default=0),
            'high_priority_files': sum(1 for f in file_metas if f.get('priority_score', 0) > 100)
        },
        'source_analysis': {
            'watch_directories': list(set(f.get('watch_dir', 'unknown') for f in file_metas)),
            'directory_file_counts': {},
            'sample_ids': list(set(f.get('sample_id', 'unknown') for f in file_metas))
        },
        'timing_analysis': {
            'batch_creation_time_ms': batch_meta.get('batch_timestamp', 0),
            'average_file_age_ms': round(sum(f.get('file_age_ms', 0) for f in file_metas) / len(file_metas), 2) if file_metas else 0,
            'oldest_file_age_ms': max((f.get('file_age_ms', 0) for f in file_metas), default=0),
            'newest_file_age_ms': min((f.get('file_age_ms', 0) for f in file_metas), default=0)
        },
        'performance_metrics': {
            'files_per_second': 0,  # Will be calculated in cumulative stats
            'mb_per_second': 0,     # Will be calculated in cumulative stats
            'reads_per_second': 0   # Will be calculated in cumulative stats
        }
    }
    
    # Calculate directory-specific file counts
    for file_meta in file_metas:
        watch_dir = file_meta.get('watch_dir', 'unknown')
        if watch_dir not in snapshot_stats['source_analysis']['directory_file_counts']:
            snapshot_stats['source_analysis']['directory_file_counts'][watch_dir] = 0
        snapshot_stats['source_analysis']['directory_file_counts'][watch_dir] += 1
    
    # Add quality indicators
    snapshot_stats['quality_indicators'] = {
        'large_files_ratio': sum(1 for f in file_metas if f.get('file_size', 0) > 50_000_000) / len(file_metas) if file_metas else 0,
        'compressed_ratio': snapshot_stats['file_statistics']['compressed_files'] / len(file_metas) if file_metas else 0,
        'high_priority_ratio': snapshot_stats['priority_analysis']['high_priority_files'] / len(file_metas) if file_metas else 0
    }
    
    # Save snapshot statistics
    output_file = "${batch_meta.batch_id}_snapshot.json"
    with open(output_file, 'w') as f:
        json.dump(snapshot_stats, f, indent=2)
    
    print(f"Generated snapshot statistics for batch {batch_meta['batch_id']}")
    print(f"Files processed: {len(file_metas)}")
    print(f"Total size: {snapshot_stats['file_statistics']['total_size_mb']} MB")
    print(f"Estimated reads: {snapshot_stats['file_statistics']['estimated_total_reads']:,}")
    
    # Generate versions file
    with open('versions.yml', 'w') as f:
        f.write('''"${task.process}":
    python: "3.9"
    statistics_framework: "1.0"''')
    """

    stub:
    """
    echo '{"batch_id": "${batch_meta.batch_id}", "stub": true}' > ${batch_meta.batch_id}_snapshot.json
    
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: "3.9"
        statistics_framework: "1.0"
    END_VERSIONS
    """
}