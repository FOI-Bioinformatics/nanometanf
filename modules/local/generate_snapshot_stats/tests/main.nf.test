nextflow_process {

    name "Test GENERATE_SNAPSHOT_STATS"
    script "../main.nf"
    process "GENERATE_SNAPSHOT_STATS"

    tag "modules"
    tag "modules_local"
    tag "generate_snapshot_stats"

    test("Should generate basic snapshot statistics") {

        tag "basic_snapshot"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        batch_id: 'batch_001',
                        batch_timestamp: 1640995200000,
                        batch_time: '2022-01-01T00:00:00Z'
                    ],
                    [
                        [
                            file_name: 'sample1.fastq.gz',
                            file_path: '/data/sample1.fastq.gz',
                            file_size: 52428800,
                            estimated_reads: 10000,
                            is_compressed: true,
                            priority_score: 85,
                            watch_dir: '/data/run1',
                            sample_id: 'sample_1',
                            file_age_ms: 120000
                        ],
                        [
                            file_name: 'sample2.fastq.gz',
                            file_path: '/data/sample2.fastq.gz',
                            file_size: 41943040,
                            estimated_reads: 8000,
                            is_compressed: true,
                            priority_score: 75,
                            watch_dir: '/data/run1',
                            sample_id: 'sample_2',
                            file_age_ms: 90000
                        ]
                    ]
                ]
                input[1] = [
                    enable_quality_analysis: true,
                    priority_threshold: 80
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should handle empty batch gracefully") {

        tag "empty_batch"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        batch_id: 'batch_empty',
                        batch_timestamp: 1640998800000,
                        batch_time: '2022-01-01T01:00:00Z'
                    ],
                    []
                ]
                input[1] = [
                    handle_empty_batches: true,
                    generate_placeholder_stats: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should handle multiple directories and high priority files") {

        tag "multi_directory"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        batch_id: 'batch_multi',
                        batch_timestamp: 1640995800000,
                        batch_time: '2022-01-01T00:10:00Z'
                    ],
                    [
                        [
                            file_name: 'urgent1.fastq.gz',
                            file_size: 62914560,
                            estimated_reads: 12000,
                            is_compressed: true,
                            priority_score: 150,
                            watch_dir: '/urgent',
                            sample_id: 'critical_case',
                            file_age_ms: 30000
                        ],
                        [
                            file_name: 'run2_sample1.fastq.gz',
                            file_size: 41943040,
                            estimated_reads: 8000,
                            is_compressed: true,
                            priority_score: 95,
                            watch_dir: '/data/run2',
                            sample_id: 'patient_B',
                            file_age_ms: 45000
                        ],
                        [
                            file_name: 'normal.fastq.gz',
                            file_size: 31457280,
                            estimated_reads: 6000,
                            is_compressed: true,
                            priority_score: 80,
                            watch_dir: '/standard',
                            sample_id: 'routine_sample',
                            file_age_ms: 180000
                        ]
                    ]
                ]
                input[1] = [
                    priority_analysis: true,
                    high_priority_threshold: 100
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
