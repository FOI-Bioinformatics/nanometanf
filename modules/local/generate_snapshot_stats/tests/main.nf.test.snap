{
    "Should handle multiple directories and high priority files": {
        "content": [
            {
                "0": [
                    [
                        {
                            "batch_id": "batch_multi",
                            "batch_timestamp": 1640995800000,
                            "batch_time": "2022-01-01T00:10:00Z"
                        },
                        "batch_multi_snapshot.json:md5,26d9060a1b0b4f9060845b0fe3c20f04"
                    ]
                ],
                "1": [
                    "versions.yml:md5,c25402591477fbd4e217997985bd994f"
                ],
                "snapshot_stats": [
                    [
                        {
                            "batch_id": "batch_multi",
                            "batch_timestamp": 1640995800000,
                            "batch_time": "2022-01-01T00:10:00Z"
                        },
                        "batch_multi_snapshot.json:md5,26d9060a1b0b4f9060845b0fe3c20f04"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c25402591477fbd4e217997985bd994f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T09:21:37.386107"
    },
    "Should generate basic snapshot statistics": {
        "content": [
            {
                "0": [
                    [
                        {
                            "batch_id": "batch_001",
                            "batch_timestamp": 1640995200000,
                            "batch_time": "2022-01-01T00:00:00Z"
                        },
                        "batch_001_snapshot.json:md5,cce29b9398f7544d6d4e5e1e0a3d624f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,c25402591477fbd4e217997985bd994f"
                ],
                "snapshot_stats": [
                    [
                        {
                            "batch_id": "batch_001",
                            "batch_timestamp": 1640995200000,
                            "batch_time": "2022-01-01T00:00:00Z"
                        },
                        "batch_001_snapshot.json:md5,cce29b9398f7544d6d4e5e1e0a3d624f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c25402591477fbd4e217997985bd994f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T09:21:29.908353"
    },
    "Should handle empty batch gracefully": {
        "content": [
            {
                "0": [
                    [
                        {
                            "batch_id": "batch_empty",
                            "batch_timestamp": 1640998800000,
                            "batch_time": "2022-01-01T01:00:00Z"
                        },
                        "batch_empty_snapshot.json:md5,478e16eb52ca34cfc50726b162aa3936"
                    ]
                ],
                "1": [
                    "versions.yml:md5,c25402591477fbd4e217997985bd994f"
                ],
                "snapshot_stats": [
                    [
                        {
                            "batch_id": "batch_empty",
                            "batch_timestamp": 1640998800000,
                            "batch_time": "2022-01-01T01:00:00Z"
                        },
                        "batch_empty_snapshot.json:md5,478e16eb52ca34cfc50726b162aa3936"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c25402591477fbd4e217997985bd994f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T09:21:33.825701"
    }
}